README
SHED (Simple Hex EDitor) v1.16

Shed is a program for viewing and editing non-text files. The look and feel is
designed to be similar to that of the pico editor.

LICENSE
Shed is released under the GNU General Public License (GPL). See COPYING for
more information.

USAGE
****************************************************
BACKUP YOUR FILES FIRST!!!.
NB: Notice there is no save key. This is because the file is not loaded into
memory, so when you edit it, the change is made directly to disk.
****************************************************

Type shed and a filename. If you're editing a file it must be an existing
file. If no filename is given, or it is '-', then shed displays stdin.

The man page has information on command line options.

Each line of the shed display represents one byte of the file. Each lines
consists of the offset of that byte from the start of the file, the ascii
value of the byte, and then the bytes value in various bases. All values are
unsigned.

Keys:

arrow keys       - move cursor around
pgup/^Y          - cursor up 16
pgdown/^V        - cursor down 16
home/^A          - start of line
end/^E           - end of line

space/e          - edit value of current byte.
                   You are asked to enter the new value in the format of the
                   current column the cursor is in - eg. if you are in the hex
                   column, it will ask for the new value in hex.

1,2,4            - set size of cursor in bytes, in the hex/dec/oct columns (and
                   binary if in traditional mode).

`                - toggle the endianness of multi-byte values (when cursor size
                   is 2 or 4)

s, w, ^x, f, F3  - search. Search string is asked for in the format of the
                   current cursor column. In ascii, just enter the string in
                   ascii. In the others, enter the values seperated by spaces -
                   eg to search for ABC from the decimal column you would enter
                   "65 66 67" (without the quotes).

^f,^b            - search shortcut for forwards/backwards respectivly.

r,n,F3           - repeat previous search

t                - toggle whether offset numbers and cursor position etc are in
                   dec or hex.

b                - toggle binary mode column behaviour (bit editing mode vs
                   traditional)

d                - dump to file.

j                - jump to byte. enter a byte number, or 'top' or 'end'.

a                - change the mode of the ascii column (printable only / c -
                   style extended chars / 'man ascii' descriptions)

p                - turn on 'preview' mode.

x, ^x            - exit


When a command (except search) asks you for input, entering a zero-length
string will cancel the command, as will ^C. When searching, entering a zero-
length string will use the previous search word, if one exists.

NOTES
shed can be used directly on devices (/dev/*) but it can not obtain the length
so you can overrun the length of the device.

CONTACT
Visit the shed homepage at http://shed.sourceforge.net

Send comments/bug reports/improvements to alexsisson@gmail.com - Please put
shed in the subject so it gets filtered to my shed folder!


ChangeLog
1.16
  - --dump now works with input from stdin
  - --hlo

1.15
  - file type descriptions
  - multi-byte editing (press 2 or 4)
  - case insensitive search
  - better backspace support
  - improved large-file support
  - units (K,M,G) support for jump command
  - -L / --length option
  - -n / --nofollow if --enable-nofollow configure option used (pretty useless)

1.14
  - experimental largefile support
  - --hex command line option
  - little interface fix for cygwin

1.13
  - binary editing (bit toggling)
  - minor interface improvments

1.12
  - read from stdin
  - fix for searching for values greater than 127 (duh!)
  - ^W added for search to be like pico (thought it was like that already)
  - ^H added for deleting input (should I be looking up the configured
    key from the env?)
  - minor fix in spacing
  - fixed bug when doing a repeat search by entering an empty string.
  - dump progress
  - --start fix

1.11
  - ^c handling (you can cancel when asked for input)
  - ^o save message :-)
  - --readonly option
  - --start option
  - preview mode (nice idea, but came out ugly, but have left it in anyway)
  - added extended ascii character display (also ugly!)
  - interface tidying
  - new keypresses!
  - more beeping on errors (can't remember what specifically)
  - ported to c
  - bit of code refactoring
  - added proper 'missing', 'install-sh' and 'mkinstalldirs' needed for RPM
    making. (not that I actually make one, but in case you do. Thanks to
    Dennis Bjorklund for info.

1.10
  - Fixed serious dump bug, where it would dump to the input file, overwriting
    its contents as it went.

1.09
  - Compiling with g++ 3.2 gave me a good screenful or two about some
    iostream deprectation, so all IO is now done with stdio (probably
    faster anyway)
  - Fixed bug with files whose size could not be obtained (devices etc).
  - Ignoring of ^C to prevent accidental quitting (see TODO)
  - Added configure script.

1.08
  - Added "repeat previous search" on R
  - Added handling of invalid digits when inputting numbers
  - Added checking for out-of-range inputs

1.07
  - Fixed some major search bugs (Thanks to Keith Williams for pointing them
    out)
  - Added ^L refresh
  - Added a very simple man page

1.06
  - Now a previous search string for each column
  - Improved error message when file could not be 'stat'ed
  - Pressing 'f' (find) now does search too

1.05
  - Fixed bug with toggling dec/hex mode on large files
  - Now checks if the argument is a directory. Before it would try and open
    the directory as a file and wouldn't do much.
  - On files where the file size could not be obtained (devices etc). offset
    column is now set to big

1.04
  - Made jump-to code more efficient - On large files it was taking a couple of
    seconds to jump from start to end

1.03
  - Fixed problem where screen wouldn't be redrawn after a resize
  - Tidied up redraw when resize takes place with cursor near EOF
  - Improved makefile's make install option. (I was using cp and chmod because
    I didn't know the install command :-) )

1.02
  - Fixed bug when using jump to end

1.01
  - Added <stdlib.h> for similar reasons as below

1.00
  - Same as 0.92, but decided it was ready for 1.00, as I can't
    be bothered to do any of the ideas at the moment. Only change
    was adding <string.h> to includes, because on my system it was
    getting included through another header. I tried building on
    another system and string.h wasn't getting included so needed
    to be explicitly included.

0.92
  - Support for read-only files.
  - Beeping when cursor is at a boundary, or when trying to
    edit a readonly file

0.91
  - Search strings can now be entered in any of the bases

0.90
  - First release to celebrate(?) it getting on sourceforge



enc.c
/*
$Id: enc.c,v 1.1 2011/01/08 15:26:50 alexsisson Exp $

(C) Copyright 2010 Alex Sisson (alexsisson@gmail.com)

Functions for decoding multi-byte character encodings.

This file is part of shed.

shed is free software; you can redistribute it and/or modify
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

shed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with shed; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <wchar.h>
#include <stdio.h>

wchar_t getmbchar_utf8(unsigned char c1, FILE *f, int *width) {
  int i,c2,c3,c4;
  //wchar_t r;
  (*width) = 0;

  if((c1 & 0x80)==0)
    return (wchar_t)c1;

  if((c1 & 0x40)==0)
    return 0; // shouldn't occur

  (*width)++;
  c2 = fgetc(f);
  if(c2<0)
    return 0; // file corrupted?

  if((c2 & 0x20)==0)
    return (((wchar_t)(c1&0x1F))<<0x06) | ((wchar_t)(c2&0x3F));

  (*width)++;
  c3 = fgetc(f);
  if(c3<0)
    return (((wchar_t)(c1&0x0F))<<0x0C) | (((wchar_t)(c2&0x3F))<<6) | ((wchar_t)(c3&0x3F));

  return 0;
}


enc.h
/*
$Id: enc.h,v 1.1 2011/01/08 15:26:50 alexsisson Exp $

(C) Copyright 2010 Alex Sisson (alexsisson@gmail.com)

Functions for handling various character encodings.

This file is part of shed.

shed is free software; you can redistribute it and/or modify
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

shed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with shed; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <wchar.h>

wchar_t getmbchar_utf8(unsigned char c1, FILE *f, int *width);



util.h
/*
$Id: util.h,v 1.7 2011/01/12 01:25:46 alexsisson Exp $

(C) Copyright 2002-2003 Alex Sisson (alexsisson@gmail.com)

This file is part of shed.

shed is free software; you can redistribute it and/or modify
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

shed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with shed; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#ifndef UTIL_H_INCLUDED
#define UTIL_H_INCLUDED

#if SHED_WIN32 != 1
#include <getopt.h>
#endif
#include <stdint.h>

/* defines */
#define strequ(s1,s2) (!(strcmp(s1,s2)))

/* ascii control char descs */
extern char  ascii_short_desc[];
extern char *ascii_long_desc[];

/* functions */
char *getascii(unsigned char c, char *s, int mode);
char *getstring(uint64_t n, char *s, int base, int width);
int64_t parsestring(char *s, unsigned int base);
int calcwidth(uint64_t n, unsigned int base);
char *getopt_makeoptstring(struct option *opt);

#endif


win32/win32.c
/*
$Id: win32.c,v 1.6 2011/06/20 00:41:27 alexsisson Exp $

(C) Copyright 2011 Alex Sisson (alexsisson@gmail.com)

This file is part of shed.

shed is free software; you can redistribute it and/or modify
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

shed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with shed; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <Windows.h>
#include <stdio.h>
#include <stdarg.h>
#include "win32.h"

static char debug[256];

WINDOW	*stdscr = NULL;
int		COLS = 0;
int		LINES = 0;

static HWND sc_hwnd = NULL;
static HDC  sc_hdc  = NULL;
static int	CW = 0; /* char width  */
static int	CH = 0; /* char height */

static char sc_str_buffer[256];
static int    kb[32], *kp = kb; /* key buffer, and pointer used for getch / ungetch */
static MEVENT mb[32], *mp = mb; /* mouse event buffer and pointer */

static unsigned long color[8];

void sc_set_hwnd(HWND hwnd) {
	sc_hwnd = hwnd;
}

void sc_set_hdc(HDC hdc) {
	HFONT		hfont, holdfont;
	RECT		r;
	TEXTMETRIC	tm;
	double		d;

	sc_hdc		= hdc;
	if(hdc) {
		hfont		= (HFONT)GetStockObject(ANSI_FIXED_FONT); 
		holdfont	= (HFONT)SelectObject(hdc,hfont);
		GetClientRect(sc_hwnd,&r);
		GetTextMetricsW(hdc,&tm);
		CW = tm.tmMaxCharWidth;
		CH = tm.tmHeight;
		d = r.bottom - r.top;
		d /= CH;
		LINES = (int)d;
		d = r.right - r.left;
		d /= CW;
		COLS = (int)d;
	}
}

int peek(void) {
	return (kb==kp) ? -1 : *kp;
}

int ungetchvk(WPARAM vk) {
	OutputDebugStringA("ungetchvk: begin\n");
	switch(vk) {
		case VK_UP:
		case VK_DOWN:
		case VK_LEFT:
		case VK_RIGHT:
		case VK_PRIOR:
		case VK_NEXT:
		case VK_BACK:
		case VK_HOME:
		case VK_END:
		case VK_F1:
		case VK_F2:
		case VK_F3:
		case VK_F4:
		case VK_F5:
		case VK_F6:
		case VK_F7:
		case VK_F8:
		case VK_F9:
		case VK_F10:
		case VK_F11:
		case VK_F12:
			ungetch(vk+KEY_NUDGE);
			break;
		default:
			return 1;
/*
		case VK_SHIFT:
			OutputDebugStringA("ungetchvk: shift (ignored)\n");
			break;
		case VK_CONTROL:
			OutputDebugStringA("ungetchvk: ctrl (ignored)\n");
			break;
		default:
			if(isdigit(vk)||isspace(vk)) {
				OutputDebugStringA("ungetchvk: digit/space\n");
				ungetch(vk);
			} else if(isupper(vk)) {
				if(GetKeyState(VK_SHIFT)&0x8000) {
					OutputDebugStringA("ungetchvk: shift+alpha\n");
					ungetch(GetKeyState(VK_CAPITAL)&0x01?tolower(vk):vk);
				} else if(GetKeyState(VK_CONTROL)&0x8000) {
					OutputDebugStringA("ungetchvk: ctrl+alpha\n");
					ungetch(vk - 64);
				} else {
					OutputDebugStringA("ungetchvk: alpha\n");
					ungetch(GetKeyState(VK_CAPITAL)&0x01?vk:tolower(vk));
				}
			} else switch(vk) {
				case VK_CAPITAL:
					return 1;
				default:
					OutputDebugStringA("ungetchvk: special\n");
					ungetch(KEY_NUDGE + vk);
					break;
			}
			break;
*/
	}
	OutputDebugStringA("ungetchvk: end\n");
	return 0;
}

/* ncurses functions */

WINDOW *initscr() {
	stdscr = (WINDOW*)calloc(1,sizeof(WINDOW));
	color[COLOR_BLACK]   = RGB(0x00,0x00,0x00);
	color[COLOR_RED]     = RGB(0xFF,0x00,0x00);
	color[COLOR_GREEN]   = RGB(0x00,0xFF,0x00);
	color[COLOR_YELLOW]  = RGB(0xFF,0xFF,0x00);
	color[COLOR_BLUE]    = RGB(0x00,0x00,0xFF);
	color[COLOR_MAGENTA] = RGB(0xFF,0x00,0xFF);
	color[COLOR_CYAN]    = RGB(0x00,0xFF,0xFF);
	color[COLOR_WHITE]   = RGB(0xFF,0xFF,0xFF);
	return stdscr;
}

int start_color() {
	return 0;
}

int init_pair(short pair, short f, short b) {
	return 0;
}

int getch(void) {
	return (kb==kp) ? ERR : *(--kp);
}

int ungetch(int ch) {
	*(kp++) = ch;
	return OK;
}

int getmouse(MEVENT *event) {
	double d;
	memset(event,0,sizeof(MEVENT));
	if(mb==mp)
		return ERR;
	memcpy(event,--mp,sizeof(MEVENT));
	d = event->x;
	d /= CW;
	event->x = d;
	d = event->y;
	d /= CH;
	event->y = d;
	return OK;
}

int ungetmouse(MEVENT *event) {
	memcpy(mp++,event,sizeof(MEVENT));
	return OK;
}

int move(int y, int x) {
	stdscr->_curx = x;
	stdscr->_cury = y;
	return OK;
}

int beep(void) {
	return ERR;
}

int refresh(void) {

}

int erase(void) {

}

int attron(int attr) {
	switch(attr) {
		case A_REVERSE:
			SetTextColor(sc_hdc, color[COLOR_BLACK]); SetBkColor(sc_hdc, color[COLOR_WHITE]);	break;

		default:
			/* else assume COLOR_PAIRs */
			/* warning this is a massive hack as we don't actually have a color pair array ! */
			SetTextColor(sc_hdc, color[attr]); SetBkColor(sc_hdc, color[COLOR_BLACK]);	break;


	}
	return OK;
}

int attroff(int attr) {
	switch(attr) {
		case A_REVERSE:
			SetTextColor(sc_hdc, color[COLOR_WHITE]); SetBkColor  (sc_hdc, color[COLOR_BLACK]);	break;
		default:
			SetTextColor(sc_hdc, color[COLOR_WHITE]); SetBkColor  (sc_hdc, color[COLOR_BLACK]);	break;
	}
	return OK;
}

int curs_set(int f) {

}

int addch(int ch) {
	sc_str_buffer[0] = ch;
	sc_str_buffer[1] = '\0';
	addstr(sc_str_buffer);
	return OK;
}

int mvaddch(int y, int x, int ch) {
	move(y,x);
	addch(ch);
	return OK;
}

int addstr(const char *str) {
	int len = strlen(str);
	int release = 0;
	if(!sc_hdc) {
		sc_hdc = GetDC(sc_hwnd);
		if(!sc_hdc) {
			return -1;
		}
		release = 1;
	}
	TextOutA(sc_hdc,stdscr->_curx*CW,stdscr->_cury*CH,str,len);
	stdscr->_curx += len;
	if(release) {
		ReleaseDC(sc_hwnd,sc_hdc);
		sc_hdc = NULL;
	}
	return OK;
}

int mvaddstr(int y, int x, const char *str) {
	move(y,x);
	addstr(str);
	return OK;
}

int printw(const char *fmt, ...) {
	va_list va;
	va_start(va,fmt);
	vsprintf(sc_str_buffer,fmt,va);
	va_end(va);
	addstr(sc_str_buffer);
	return OK;
}

int mvprintw(int y, int x, const char *fmt, ...) {
	va_list va;
	va_start(va,fmt);
	vsprintf(sc_str_buffer,fmt,va);
	va_end(va);
	mvaddstr(y,x,sc_str_buffer);
	return OK;
}

int addwstr(const wchar_t *wstr) {

}


win32/win32.h
/*
$Id: win32.h,v 1.4 2011/06/20 00:41:27 alexsisson Exp $

(C) Copyright 2011 Alex Sisson (alexsisson@gmail.com)

This file is part of shed.

shed is free software; you can redistribute it and/or modify
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

shed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with shed; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

===============================================================

This is essentially a bespoke implementation of a tiny subset of curses,
providing just enough features to get shed working in a Windows window with-
out major changes to the shed code itself.

I've kept it non-shed specific to allow for potential reuse, and possible
expansion to be more curses-compliant. However this may pose some challenges
(trying to implement non-event driven curses into the event-driven world of
win32)

*/

#include <Windows.h>

#define OK				0
#define ERR				(-1)

/* map win32 key constants to ncurses constants */
#define KEY_NUDGE		0xff   /* nudge special characters up above the 0xff so they don't clash with chars */
#define KEY_MOUSE		0xffff

#define KEY_UP			(KEY_NUDGE + VK_UP)
#define KEY_DOWN		(KEY_NUDGE + VK_DOWN)
#define KEY_LEFT		(KEY_NUDGE + VK_LEFT)
#define KEY_RIGHT		(KEY_NUDGE + VK_RIGHT)
#define KEY_PPAGE		(KEY_NUDGE + VK_PRIOR)
#define KEY_NPAGE		(KEY_NUDGE + VK_NEXT)
#define KEY_BACKSPACE	(KEY_NUDGE + VK_BACK)
#define KEY_HOME		(KEY_NUDGE + VK_HOME)
#define KEY_END			(KEY_NUDGE + VK_END)
#define KEY_F(N)		(KEY_NUDGE + VK_F1 + ((N)-1))
#define KEY_RESIZE		9000

#define COLOR_PAIR(N)	(N)
#define A_REVERSE		20

#define COLOR_BLACK   0
#define COLOR_RED     1
#define COLOR_GREEN   2
#define COLOR_YELLOW  3
#define COLOR_BLUE    4
#define COLOR_MAGENTA 5
#define COLOR_CYAN    6
#define COLOR_WHITE   7

#define BUTTON1_PRESSED 1

extern int COLS;
extern int LINES;

struct window {
	int _curx, _cury;
};
typedef struct window WINDOW;
extern WINDOW *stdscr;

struct mevent {
	int x, y;
	unsigned long bstate;
};
typedef struct mevent MEVENT;

/* functions related to our crude curses "implementation" */
void sc_set_hwnd(HWND hwnd);
void sc_set_hdc(HDC hdc);
int peek(void);
int ungetchvk(WPARAM vk);

/* curses functions */

WINDOW *initscr();

int start_color(void);
int init_pair(short pair, short f, short b);

int getch(void);
int ungetch(int ch);

int getmouse(MEVENT *event);
int ungetmouse(MEVENT *event);

#define getyx(W,Y,X)	(Y)=(W)->_cury;(X)=(W)->_curx;

int move(int y, int x);
int beep(void);
int refresh(void);
int erase(void);

int attron(int attr);
int attroff(int attr);

int curs_set(int);

int addch(int ch);
int mvaddch(int y, int x, int ch);

int addstr(const char *str);
int mvaddstr(int y, int x, const char *str);

int printw(const char *fmt, ...);
int mvprintw(int y, int x, const char *fmt, ...);

int addwstr(const wchar_t *wstr);



win32/Resource.h
//{{NO_DEPENDENCIES}}
// Microsoft Visual C++ generated include file.
// Used by shed.rc
//
#define IDC_MYICON                      2
#define IDD_SHED_DIALOG                 102
#define IDS_APP_TITLE                   103
#define IDD_ABOUTBOX                    103
#define IDM_ABOUT                       104
#define IDM_EXIT                        105
#define IDI_SHED                        107
#define IDI_SMALL                       108
#define IDC_SHED                        109
#define IDR_MAINFRAME                   128
#define IDD_GETINPUT                    129
#define IDC_GETINPUT_STATIC             1000
#define IDC_EDIT1                       1001
#define IDC_GETINPUT_EDIT               1001
#define ID_FILE_OPEN                    32771
#define IDM_FILE_OPEN                   32772
#define IDM_FILE_EXIT                   32773
#define IDC_STATIC                      -1

// Next default values for new objects
// 
#ifdef APSTUDIO_INVOKED
#ifndef APSTUDIO_READONLY_SYMBOLS
#define _APS_NO_MFC                     1
#define _APS_NEXT_RESOURCE_VALUE        130
#define _APS_NEXT_COMMAND_VALUE         32774
#define _APS_NEXT_CONTROL_VALUE         1002
#define _APS_NEXT_SYMED_VALUE           110
#endif
#endif


util.c
/*
$Id: util.c,v 1.10 2011/01/12 01:25:46 alexsisson Exp $

(C) Copyright 2002-2004 Alex Sisson (alexsisson@gmail.com)

This file is part of shed.

shed is free software; you can redistribute it and/or modify
under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

shed is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with shed; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

/* includes */
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include <stdint.h>

#include "util.h"

/* ascii control char descs */
char  ascii_short_desc[] = {'0',0,0,0,0,0,0,'a','b','t','n','v','f','r'};
char *ascii_long_desc[]  = {"NUL","SOH","STX","ETX","EOT","ENQ","ACK","BEL",
                            "BS ","HT ","LF ","VT ","FF ","CR ","SO ","SI ",
                            "DLE","DC1","DC2","DC3","DC4","NAK","SYN","ETB",
                            "CAN","EM ","SUB","ESC","FS ","GS ","RS ","US ","SPC"};



/* returns a description of the char if its <32 || >126 */
char *getascii(unsigned char c, char *s, int mode)
{
  if(c>127)
    strncpy(s,"   ",3);
  else if(c>=32 && c<=126) {
    s[0] = ' ';
    s[1] = c;
    s[2] = ' ';
  } else {
    switch(mode) {
      case 1: /* do c-style control chars */
        if(c!=127 && ascii_short_desc[c]) {
          s[0] = ' ';
          s[1] = '\\';
          s[2] = ascii_short_desc[c];
          break;
        }

      case 0: /* no special chars */
        strncpy(s,"   ",3);
        break;

      case 2: /* 3 letter descs */
        strncpy(s,c!=127?ascii_long_desc[c]:"DEL",3);
        break;
    }
  }
  s[3] = 0;
  return s;
}

/* returns string representation of n in base 'base' */
char *getstring(uint64_t n, char *s, int base, int width) {
  uint64_t i,j,c,d = calcwidth(n,base);
  char *p = s;
  if(width && d<width) {
    for(i=d;i<width;i++) {
      *p = '0';
      p++;
    }
  }
  for(i=d;i;i--) {
    j = pow(base,i-1);
    c = 0;
    while(n>=j) {
      n-=j;
      c++;
    }
    *p = c + ((c<10) ? 48 : 55);
    p++;
  }
  *p = 0;
  return s;
}

/* parses 's' as a 'base' base string */
int64_t parsestring(char *s, unsigned int base) {

  unsigned int i,j,slen=strlen(s),n=0;
  int64_t r = 0;

  for(i=0;i<slen;i++) {
    if(s[i]>='0' && s[i]<='9')
      n = s[i] - 48;
    else if(s[i]>='A' && s[i]<='F')
      n = s[i] - 55;
    else if(s[i]>='a' && s[i]<='f')
      n = s[i] - 87;
    else
      return -1;
    if(n>=base)
      return -1;
    for(j=1;j<slen-i;j++)
      n*=base;
    r+=n;
  }
  return r;
}

#include <stdio.h>

/* calculates the width needed to represent 'n' in base 'base' */
int calcwidth(uint64_t n, unsigned int base) {
  unsigned int i,j;
  uint64_t b;
  if(!base)
    return 0;
  for(i=0;;i++) {
    b = base;
    for(j=0;j<i;j++)
      b *= base;
    if(b>n || b<base)
      return i + 1;
  }
  return 0;
}

#if SHED_WIN32 != 1
/* creates optstring arg (short opts) from for getopt */
char *getopt_makeoptstring(struct option *opt) {
  int n = 0;
  char *optstring,*p;
  struct option *o = opt;

  while(o->name || o->has_arg || o->flag || o->val) {
    n += 1 + o->has_arg;
    if(o->flag)
      return NULL; /* fail because not returning o->val */
    o++;
  }

  p = optstring = malloc(n+1);
  o = opt;

  while(o->name || o->has_arg || o->flag || o->val) {
    *p++ = o->val;
    n = o->has_arg;
    while(n--)
      *p++ = ':';
    o++;
  }
  *p = 0;

  return optstring;
}

#endif